File: sorting_example2.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

struct Student {
  std::string name;
  int grade;
};

// Συνάρτηση για εμφάνιση της λίστας μαθητών
void print_students(const std::vector<Student>& students) {
  for (const auto& student : students) {
    std::cout << student.name << " - " << student.grade << "\n";
  }
}

int main() {
  std::vector<Student> students = {
      {"Alice", 85}, {"Bob", 92}, {"Charlie", 78}, {"David", 90}};

  std::cout << "Πριν την ταξινόμηση:\n";
  print_students(students);

  // Ταξινόμηση κατά βαθμό (φθίνουσα σειρά)
  std::sort(
      students.begin(), students.end(),
      [](const Student& a, const Student& b) { return a.grade < b.grade; });

  std::cout << "\nΜετά την ταξινόμηση κατά βαθμό:\n";
  print_students(students);

  // Ταξινόμηση κατά όνομα (αλφαβητικά σε φθίνουσα σειρά)
  std::sort(students.begin(), students.end(),
            [](const Student& a, const Student& b) { return a.name > b.name; });

  std::cout << "\nΜετά την ταξινόμηση κατά όνομα (αλφαβητικά σε φθίνουσα σειρά):\n";
  print_students(students);

  return 0;
}

File: SortingExample2.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

class Student {
    String name;
    int grade;

    Student(String name, int grade) {
        this.name = name;
        this.grade = grade;
    }

    @Override
    public String toString() {
        return name + " - " + grade;
    }
}

public class SortingExample2 {
    public static void main(String[] args) {
        List<Student> students = new ArrayList<>();
        students.add(new Student("Alice", 85));
        students.add(new Student("Bob", 92));
        students.add(new Student("Charlie", 78));
        students.add(new Student("David", 90));

        System.out.println("Πριν την ταξινόμηση:");
        for (Student student : students) {
            System.out.println(student);
        }

        // Ταξινόμηση κατά βαθμό (φθίνουσα σειρά)
        students.sort(Comparator.comparingInt((Student s) -> s.grade));
        System.out.println("\nΜετά την ταξινόμηση κατά βαθμό:");
        for (Student student : students) {
            System.out.println(student);
        }

        // Ταξινόμηση κατά όνομα (φθίνουσα αλφαβητική σειρά)
        students.sort(Comparator.comparing((Student s) -> s.name).reversed());
        System.out.println("\nΜετά την ταξινόμηση κατά όνομα (φθίνουσα):");
        for (Student student : students) {
            System.out.println(student);
        }
    }
}
